src/org/jetbrains/plugins/scala/compiler/ScalaHydraCompilerConfigurationPanel.scala

```diff
@@ -0,0 +1,74 @@
+package org.jetbrains.plugins.scala.compiler
+
+import java.awt.event.{ActionEvent, FocusEvent, FocusListener}
+import javax.swing.event.DocumentEvent
+
+import com.intellij.openapi.ui.Messages
+import org.jetbrains.plugins.scala.caches.HydraArtifactsCache
+import org.jetbrains.plugins.scala.extensions
+import com.intellij.openapi.project.Project
+import com.intellij.ui.DocumentAdapter
+import org.jetbrains.plugins.scala.project.{Platform, ProjectExt, Version, Versions}
+
+import scala.collection.JavaConverters._
+import scala.util.{Failure, Success}
+
+/**
+  * @author Maris Alexandru
+  */
+class ScalaHydraCompilerConfigurationPanel(project: Project, settings: HydraCompilerSettings) extends HydraCompilerConfigurationPanel {
+  private val UnknownVersion = "unknown"
+
+  val documentAdapter = new DocumentAdapter {
+      override def textChanged(documentEvent: DocumentEvent): Unit =
+        if (!getUsername.isEmpty && !getPassword.isEmpty) downloadButton.setEnabled(true)
+        else downloadButton.setEnabled(false)
+    }
+
+  val focusListener = new FocusListener {
+    override def focusGained(e: FocusEvent) = {}
+
+    override def focusLost(e: FocusEvent) = if (!getUsername.isEmpty && !getPassword.isEmpty &&
+      (HydraCredentialsManager.getLogin != getUsername || HydraCredentialsManager.getPlainPassword != getPassword)) {
+      HydraCredentialsManager.setCredentials(getUsername, getPassword)
+      hydraVersionComboBox.setItems(Versions.loadScalaVersions(Platform.Hydra))
+    }
+  }
+
+  userTextField.addFocusListener(focusListener)
+  userTextField.getDocument.addDocumentListener(documentAdapter)
+  passwordTextField.getDocument.addDocumentListener(documentAdapter)
+  passwordTextField.addFocusListener(focusListener)
+  hydraVersionComboBox.setItems(Versions.loadScalaVersions(Platform.Hydra))
+  downloadButton.addActionListener((_: ActionEvent) => onDownload())
+
+  def selectedVersion: String = hydraVersionComboBox.getSelectedItem.asInstanceOf[String]
+
+  def onDownload() = {
+    downloadVersionWithProgress(project.scalaModules.map(module => module.sdk.compilerVersion.getOrElse(UnknownVersion)), selectedVersion)
+    settings.hydraVersion = selectedVersion
+  }
+
+  private def downloadVersionWithProgress(scalaVersions: Seq[String], hydraVersion: String): Unit = {
+    val filteredScalaVersions = scalaVersions.distinct.filterNot(_ == UnknownVersion)
+      .map(Version(_)).filter(_ >= Version("2.11")).map(_.presentation).filterNot(_ == "2.12.0")
```

Comment:
I find it a lot more readable to convert chained map/filter with a for-comprehension, i.e.,
```
for {
  rawVersion <- scalaVersions.distinct
  if rawVersion != UnknownVersion
  version = Version(rawVersion)
  if version >= Version("2.11") && rawVersion != "2.12.0"
} yield version
```

And, as an added benefit, by using a for-comprehension the created code is also more efficient! The reason is that the filer condition in the for-comprehension (i.e., the `if`) are converted into calls to `withFilter`, which allows to fuse operations so that you don't have to do multiple passes. (see https://stackoverflow.com/questions/19617378/withfilter-instead-of-filter for some background on the difference between filter and withFilter)